src/GitHub.Api/Git/Repository.cs

```diff
@@ -10,49 +10,128 @@ namespace GitHub.Unity
     [DebuggerDisplay("{DebuggerDisplay,nq}")]
     class Repository : IEquatable<Repository>, IRepository
     {
-        private ConfigBranch? currentBranch;
-        private IList<GitLock> currentLocks;
-        private ConfigRemote? currentRemote;
-        private GitStatus currentStatus;
-        private Dictionary<string, ConfigBranch> localBranches = new Dictionary<string, ConfigBranch>();
-        private Dictionary<string, Dictionary<string, ConfigBranch>> remoteBranches = new Dictionary<string, Dictionary<string, ConfigBranch>>();
-        private Dictionary<string, ConfigRemote> remotes;
         private IRepositoryManager repositoryManager;
-        public event Action<string> OnCurrentBranchChanged;
-        public event Action<string> OnCurrentRemoteChanged;
-        public event Action OnCurrentBranchUpdated;
-        public event Action OnLocalBranchListChanged;
-        public event Action<IEnumerable<GitLock>> OnLocksChanged;
-        public event Action OnRemoteBranchListChanged;
-        public event Action OnRepositoryInfoChanged;
-
-        public event Action<GitStatus> OnStatusChanged;
+        private ICacheContainer cacheContainer;
+        private UriString cloneUrl;
+        private string name;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="Repository"/> class.
         /// </summary>
-        /// <param name="name">The repository name.</param>
         /// <param name="localPath"></param>
-        public Repository(string name, NPath localPath)
+        /// <param name="container"></param>
+        public Repository(NPath localPath, ICacheContainer container)
         {
-            Guard.ArgumentNotNullOrWhiteSpace(name, nameof(name));
             Guard.ArgumentNotNull(localPath, nameof(localPath));
 
-            Name = name;
             LocalPath = localPath;
-            this.User = new User();
+            User = new User();
+
+            cacheContainer = container;
+            cacheContainer.CacheInvalidated += CacheContainer_OnCacheInvalidated;
+            cacheContainer.CacheUpdated += CacheContainer_OnCacheUpdated;
+        }
+
+        private void CacheContainer_OnCacheInvalidated(CacheType cacheType)
```

Comment:
All these private methods should be after public methods, having a hard time reviewing this because of so many unnecessary method moves...

Reply:
👍 

Apologies, I did a lot of hashing and rehashing